vcr/statuslist/store.go

```diff
@@ -0,0 +1,287 @@
+	Create(ctx context.Context, issuer did.DID, purpose StatusPurpose) (*credential.StatusList2021Entry, error)
+	// Revoke by adding StatusList2021Entry to the list of revocations.
+	// The credentialID is only used to allow reverse search of revocations, its issuer is NOT compared to the entry issuer.
+	// Returns types.ErrRevoked if already revoked, or types.ErrNotFound when the entry.StatusListCredential is unknown.
+	Revoke(ctx context.Context, credentialID ssi.URI, entry credential.StatusList2021Entry) error
+}
+
+func (s statusListCredentialRecord) TableName() string {
+	return "status_list_credentials"
+}
+
+// statusListCredentialRecord keeps track of the StatusListCredential issued by an issuer, and what the LastIssuedIndex is for each credential.
+type statusListCredentialRecord struct {
+	StatusListCredential string `gorm:"primaryKey"`
+	Issuer               string // did:web:example.com:iam:id
+	Page                 int    // >= 1
+	LastIssuedIndex      int    // 0 <= statusListIndex < statuslist2021.MaxBitstringIndex
+}
+
+func (s revocationRecord) TableName() string {
+	return "status_list_statuses"
+}
+
+// revocationRecord is created when a
+type revocationRecord struct {
+	StatusListCredential string `gorm:"primaryKey"`                     // https://example.com/iam/id/status/page
+	StatusListIndex      int    `gorm:"primaryKey;autoIncrement:false"` // 0 <= statusListIndex <= statuslist2021.MaxBitstringIndex
+	CredentialID         string // did:web:example.com:iam:id#unique-identifier
+	// TODO: do we want to store the timestamp the revocation was added?
+	//CreatedAt time.Time
+}
+
+var _ StatusList2021Issuer = (*sqlStore)(nil)
+
+type sqlStore struct {
+	db *gorm.DB
+}
+
+// DB creates a new Session with the provided context.
+func (s *sqlStore) DB(ctx context.Context) *gorm.DB {
+	return s.db.WithContext(ctx)
+}
+
+func NewStatusListStore(db *gorm.DB) (*sqlStore, error) {
+	return &sqlStore{db: db}, nil
+}
+
+func (s *sqlStore) CredentialSubject(ctx context.Context, issuer did.DID, page int) (*credential.StatusList2021CredentialSubject, error) {
+	statusListCredential, err := toStatusListCredential(issuer, page)
+	if err != nil {
+		return nil, err
+	}
+
+	// confirm that status list credential exists
+	// SELECT * FROM status_list_credentials WHERE status_list_credential = 'statusListCredential'
+	err = s.DB(ctx).First(new(statusListCredentialRecord), "status_list_credential = ?", statusListCredential).Error
+	if err != nil {
+		if errors.Is(err, gorm.ErrRecordNotFound) {
+			// status list credential does not exist or is not managed byt his node
+			return nil, errNotFound
+		}
+		return nil, err
+	}
+
+	// get revocations
+	var revocations []revocationRecord
+	// SELECT status_list_index FROM status_list_statuses WHERE status_list_credential = 'statusListCredential'
+	err = s.DB(ctx).Select("status_list_index").Find(&revocations, "status_list_credential = ?", statusListCredential).Error
```

Comment:
If you model the relationship in the `statusListCredentialRecord` and `revocationRecord ` structs with `gorm` tags, you can qurey them in 1 go with `Preload()`.